### PR TITLE
Add an alternative Saltpetre recipe

### PR DESCRIPTION
**Pull request by maybeserendipitous**
[feature][help wanted][input wanted]

This is my first PR and my first time coding in DM so a confirmation that I didn't burn everything down would be great.

## About the PR 

Relevant previous discussion here: https://forum.ss13.co/showthread.php?tid=18910
This adds a new method for the industrial production of Saltpetre. 
This method is non-instant and produces miasma as it occurs. The recipe is:
```
(1) Space Fungus + (1) Compost + (1) Potash + (1) Water -> (2) Saltpetre + (2) Water
```
Previously discussed, a big concern was that this would make it too easy to produce large amounts of Black Powder rapidly. This method is not only time-gated and largely restricted to Botany, it is also somewhat obvious as it produces large amounts of miasma in the process due to the "decomposition" of the compost. It is comparable to buying formula from Cargo in terms of speed.

## Why's this needed? 

This is discussed more in the linked thread. In essence, large botany crews need a way to obtain large amounts of Saltpetre without grinding Cargo to dust, and using Urine directly is too slow.

## Changelog

```changelog
(u)ohjoy
(*)Added a new non-instant recipe for Saltpetre that produces miasma as it occurs: (1) Space Fungus + (1) Compost + (1) Potash + (1) Water -> (2) Saltpetre + (2) Water
```

### CODE REVIEW

**Comment by Vlaad3** [comment]
As someone who plays botany a fair amount, I would love an alternative to piss. However, the reaction seems a bit out of line compared to the rest of our chems imo. Maybe a recipe in where vomit can be used? Since botany can produce it themselves.

**Comment by maybeserendipitous** [comment]
> As someone who plays botany a fair amount, I would love an alternative to piss. However, the reaction seems a bit out of line compared to the rest of our chems imo. Maybe a recipe in where vomit can be used? Since botany can produce it themselves.

1. Can you elaborate on how it is "out of line"?
2. Do you know if Rafflesia's miasma can be extracted? If so, I would be happy to replace the recipe with 

```
(1) Compost + (1) Potash + (1) Water + (1) Miasma -> (2) Saltpetre + (2) Water
```

Where the Miasma provides the "bacteria" to decompose the Compost.

**Comment by maybeserendipitous** [comment]
Alright, I just tested it and Miasma can't be extracted. I don't want to use vomit instead because although there are bacteria in vomit, I doubt the high acidity would be conducive to the bacteria precipitating the decay. I am open to using Space Fungus though, since in the nitrogen cycle, sometimes fungi are involved in ammonification.

**Comment by Vlaad3** [comment]
> 1. Can you elaborate on how it is "out of line"?
> 
 As far as I am aware, there are no time based reactions. Looking back at this, I wrote it early in the morning, I'll think on it longer and come back with better critique. Sorry.

**Comment by TDHooligan** [comment]
> > 1. Can you elaborate on how it is "out of line"?
> 
> As far as I am aware, there are no time based reactions. Looking back at this, I wrote it early in the morning, I'll think on it longer and come back with better critique. Sorry.

put gameplay oriented feedback the discussion thread - time based reactions are implemented with 'instant=0' and are used currently with Cryoxadone and Pyrosium.

Balance aside, the PR looks good. I think you've got it spot on.

**Comment by MetricDuck** [comment]
Have you checked how miasma generation interacts with large puddles (such as you might find on oshan)?

**Comment by maybeserendipitous** [comment]
@MetricDuck I haven't checked the miasma + puddle interactions yet, but a water cooler bottle of the recipe produces about twice the miasma of a corpse, and I doubt there'll be very many Saltpetre reactions on the station at any time, so I'd wager the lag impact would not be too bad.

What I am concerned about is a bug I discovered where miasma is produced constantly, even when the reaction should be completed. In my testing, once the reaction is started, excess reagents of the recipe will be consumed, and the reaction will proceed indefinitely. I don't know why this occurs, though it might just be on my end.

File: code/modules/chemistry/Chemistry-Recipes.dm [hunk 1]
```diff
@@ -2850,6 +2850,29 @@ datum
...
+			mix_phrase = "A putrid odor pours from the mixture as a white crystalline substance leaches into the water."
+			mix_sound = 'sound/misc/fuse.ogg'
+			
+			on_reaction(var/datum/reagents/holder, created_volume)
+				// water byproduct
+				// some nitrification processes create additional water.
+				holder.add_reagent("water", created_volume,,holder.total_temperature)
+				// disgusting
+				var/turf/location = get_turf(holder.my_atom)
+				location.fluid_react_single("miasma", created_volume, airborne = 1)
```
Comment (Tarmunora) [inline code/modules/chemistry/Chemistry-Recipes.dm:2875]: this won't work when in fluids. consider `pick(holder.covered_turf())` instead

Comment (maybeserendipitous) [reply code/modules/chemistry/Chemistry-Recipes.dm]: Should be fixed now, thank you for the feedback.

**Comment by trazodont** [comment]
I wish I checked this PR before it got merged.

I don't know if you play botany personally, but this new recipe actually does a massive disservice to botanists than an advantage of making saltpetre.

While I would absolutely love a new saltpetre recipe, this one prevents botanists from doing basic tray chemistry or fertilizing water tanks, since most use compost, potash and space fungus in trays or water tanks already.

I'm all for a new saltpetre recipe, but this one has completely messed with botany.

**Comment by flrsh** [comment]
Whoops didn't realize that, sorry! @maybeserendipitous not sure who you are on Discord, but I'm temporarily changing this recipe to your original suggestion of nitrogen + compost + potash; let me know what alternative recipe you decide on!

**Comment by maybeserendipitous** [comment]
> I wish I checked this PR before it got merged.
> 
> I don't know if you play botany personally, but this new recipe actually does a massive disservice to botanists than an advantage of making saltpetre.
> 
> While I would absolutely love a new saltpetre recipe, this one prevents botanists from doing basic tray chemistry or fertilizing water tanks, since most use compost, potash and space fungus in trays or water tanks already.
> 
> I'm all for a new saltpetre recipe, but this one has completely messed with botany.

I can't believe that I missed the tray chemistry interaction when designing this. Thanks for bringing it up. Major facepalm, and a good reminder to always test things in a mock live environment. 

> Whoops didn't realize that, sorry! @maybeserendipitous not sure who you are on Discord, but I'm temporarily changing this recipe to your original suggestion of nitrogen + compost + potash; let me know what alternative recipe you decide on!

`Nitrogen + Compost + Potash` is fine. Thank you for fixing my mistake so quickly. 
For readers from a darker future where players are using this to blow the station sky-high with black powder: adding `miasma` to the recipe to effectively make it Botany-only might be good enough to fix the problem.

**Pull request merged**